coalib/misc/Caching.py

```diff
@@ -0,0 +1,193 @@
+from coalib.output.Tagging import get_tags_dir as get_pickle_dir
+
+
+def get_cache_data_path(log_printer, filename):
+    """
+    Get the full path of ``filename`` present in the user's cache directory.
+
+    :param log_printer: A LogPrinter object to use for logging.
+    :param filename:    The file whose path needs to be expanded.
+    :return:            Full path of the file, assuming it's present in the
+                        user's config directory.
+    """
+    return os.path.join(get_pickle_dir(log_printer), filename)
+
+
+def delete_cache_files(log_printer, files=[caching_db, changed_files_db]):
+    """
+    Delete the cache files after displaying a warning saying the cache
+    is corrupted and will be removed.
+
+    :param log_printer: A LogPrinter object to use for logging.
+    :param files:       The list of files to be deleted.
+    :return:            True if all the given files were successfully deleted.
+                        False otherwise.
+    """
+    log_printer.warn("The caching database is corrupted and will "
+                     "be removed. Each project will be re-cached "
+                     "automatically the next time coala is run.")
+    error_files = []
+    for file_name in files:
+        file_path = get_cache_data_path(log_printer, file_name)
+        cache_dir = os.path.dirname(file_path)
+        try:
+            os.remove(file_path)
+        except OSError:
+            error_files.append(file_name)
+
+    if len(error_files) > 0:
+        error_files = ", ".join(error_files)
+        log_printer.warn("There was a problem deleting the following "
+                         "files: " + error_files + ". Please delete "
+                         "them manually from " + cache_dir)
+        return False
+
+    return True
+
+
+def pickle_load(log_printer, filename, fallback=None):
+    """
+    Get the data stored in ``filename`` present in the coala user
+    config directory. Example usage:
+
+    >>> from pyprint.NullPrinter import NullPrinter
+    >>> from coalib.output.printers.LogPrinter import LogPrinter
+    >>> log_printer = LogPrinter(NullPrinter())
+    >>> test_data = {"answer": 42}
+    >>> pickle_dump(log_printer, "test_file", test_data)
+    >>> pickle_load(log_printer, "test_file")
+    {'answer': 42}
+    >>> pickle_load(log_printer, "nonexistant_file")
+    >>> pickle_load(log_printer, "nonexistant_file", fallback=42)
+    42
+
+
+    :param log_printer: A LogPrinter object to use for logging.
+    :param filename:    The name of the file present in the coala user config
+                        directory.
+    :param fallback:    Return value to fallback to in case the file doesn't
+                        exist.
+    :return:            Data that is present in the file, if the file exists.
+                        Otherwise the ``default`` value is returned.
+    """
+    filename = get_cache_data_path(log_printer, filename)
```

Comment:
so a class should be a thing. So a class like "AnalysisState" makes sense to me, that holds all information and can be load (classmethod) from file, created (init) of course saved (method). I'm not sure if we want it to be immutable.